File: swift/CLDContext.swift
protocol CLDContext {
    var recentField: String { get set }

    func field<T>(_ name: String) -> T
    func fieldAny(_ name: String) -> Any
    mutating func setField(_ name: String, _ value: Any)
}

extension CLDContext {
    /// Default implementation of `fieldAny()` that should be enough
    /// for CLDController
    func fieldAny(_ name: String) -> Any {
        return field(name)
    }
}

File: kotlin/CLDContext.kt
interface CLDContext {
    /**
     * Name of the field that has just changed
     *
     * Allows shoulds/handlers/behaviour functions to react to only
     * relevant changes and ignore other changes of CLDContext
     */
    var recentField: String

    /**
     * Get field's value by its name
     */
    fun field(name: String): Any
    /**
     * Create a copy of the CLDContext derivative
     *
     * Used by CLDController to treat all derived contexts as CLDContext
     */
    fun selfCopy(): CLDContext
    /**
     * Set field's value by its name
     */
    fun setField(name: String, value: Any)
}
